Meta/Lagom/CMakeLists.txt

```diff
@@ -51,281 +79,459 @@ set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${LINKER_FLAGS}")
+    set(Lagom_INSTALL_CMAKEDIR "${CMAKE_INSTALL_DATADIR}/${package}"
+        CACHE PATH "CMake package config location relative to the install prefix")
+    mark_as_advanced(Lagom_INSTALL_CMAKEDIR)
+
+    install(
+        FILES ${SERENITY_PROJECT_ROOT}/Meta/CMake/lagom-install-config.cmake
+        DESTINATION "${Lagom_INSTALL_CMAKEDIR}"
+        RENAME "${package}Config.cmake"
+        COMPONENT Lagom_Development
+    )
+
+    install(
+        FILES "${PROJECT_BINARY_DIR}/${package}ConfigVersion.cmake"
+        DESTINATION "${Lagom_INSTALL_CMAKEDIR}"
+        COMPONENT Lagom_Development
+    )
+
+    install(
+        EXPORT LagomTargets
+        NAMESPACE Lagom::
+        DESTINATION "${Lagom_INSTALL_CMAKEDIR}"
+        COMPONENT Lagom_Development
+    )
+
+    # Manually install AK
+    install(DIRECTORY "${SERENITY_PROJECT_ROOT}/AK"
+        COMPONENT Lagom_Development
+        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+        FILES_MATCHING PATTERN "*.h")
 
-    include(../../Userland/Libraries/LibUnicode/unicode_data.cmake)
 endif()
 
+function(lagom_lib library fs_name)
+    cmake_parse_arguments(LAGOM_LIBRARY "" "" "SOURCES;LIBS" ${ARGN})
+    # FIXME: Consider whether to care about -DBUILD_SHARED_LIBS=OFF
+    # Possibly a cmake presets value?
+    set(target_name "Lagom${library}")
+    add_library(${target_name} SHARED ${LAGOM_LIBRARY_SOURCES})
+    # alias for pretty exports
+    add_library(Lagom::${target_name} ALIAS ${target_name})
+
+    set_target_properties(
+        ${target_name} PROPERTIES
+        VERSION "${PROJECT_VERSION}"
+        SOVERSION "${PROJECT_VERSION_MAJOR}"
+        EXPORT_NAME ${target_name}
+        OUTPUT_NAME lagom-${fs_name}
+    )
+    target_link_libraries(${target_name} ${LAGOM_LIBRARY_LIBS})
+    if (NOT ${target_name} STREQUAL "LagomCore")
+        target_link_libraries(${target_name} LagomCore)
+    endif()
+    # Don't install Lagom libs into the target Root/
+    # FIXME: Remove this check for 4594
+    if (CMAKE_SOURCE_DIR MATCHES ".*/Lagom")
+        install(
+            TARGETS ${target_name}
+            EXPORT LagomTargets
+            RUNTIME #
+                COMPONENT Lagom_Runtime
+            LIBRARY #
+                COMPONENT Lagom_Runtime
+                NAMELINK_COMPONENT Lagom_Development
+            ARCHIVE #
+                COMPONENT Lagom_Development
+            INCLUDES #
+                DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+        )
+        install(DIRECTORY "${SERENITY_PROJECT_ROOT}/Userland/Libraries/Lib${library}"
+                COMPONENT Lagom_Developmen
+                DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+                FILES_MATCHING PATTERN "*.h")
+    endif()
+endfunction()
+
+function(lagom_test source)
+    cmake_parse_arguments(LAGOM_TEST "" "" "LIBS" ${ARGN})
+    get_filename_component(name ${source} NAME_WE)
+    add_executable(${name}_lagom ${source} $<TARGET_OBJECTS:LagomTestMain>)
```

Comment:
Any reason why the `TARGET_OBJECTS` genex is used here? I don't know in which CMake version this changed, but you can [just link to object targets](https://github.com/friendlyanon/cmake-init-gif-engine/blob/master/test/CMakeLists.txt#L38) like regular targets.

Reply:
Hmm. In the serenity_test function and serenity_testjs_test functions in Meta/utils.cmake, I had (for some reason) wanted to be able to pass a custom main file as a source file to be added to the sources list.

So the TARGET_OBJECTS expression works for that case, when we want to override the "main" file in  TEST_SOURCES file with the pre-built common "main". I changed that to be the genex in a commit on this PR. Since we currently only have two different main files for LibTest tests, it might make sense to force SERENITY_TEST_CUSTOM_MAIN to be an object target and be done with it though.

Reply:
Oh so that means the reason it's like that here in Lagom/CMakeLists.txt is I copy pasted it from my earlier experiments with the serenity_test macro.

Reply:
For the record, it was [CMake 3.12](https://cmake.org/cmake/help/latest/release/3.12.html#commands) that changed how object libraries are handled. (see the last point in the linked list)